<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN">
<html>
  <head>
    <meta content="text/html; charset=utf-8" http-equiv="content-type">
    <title>RNAstructure Command Line Help: MaxExpect</title>
    <link href="Stylesheet.css" rel="stylesheet" type="text/css">
  </head>
  <body>
    <div class="bodydiv">
      <table class="mainTable" border="0" cellpadding="0" cellspacing="0">
        <tbody>
          <tr class="headerRow">
            <td class="headerBox"><img src="icon.jpg" alt="RNAstructure logo"></td>
            <td class="headerLabel">
              <h2>RNAstructure Command Line Help<br>
                MaxExpect</h2>
            </td>
            <td class="headerBox2">
              <ul id="navigator">
                <li>
                  <h3 class="noMarginOrPad"><a href="index.html">Contents</a></h3>
                </li>
                <li>
                  <h3 class="noMarginOrPad"><a href="AlphabeticalIndex.html">Index</a></h3>
                </li>
              </ul>
            </td>
          </tr>
          <tr>
            <td class="mainContent" colspan="3">
              <p>MaxExpect predicts the maximum expected accuracy structure, a
                structure that maximizes pair probabilities. The
                paiprobabilities are generated using a partition function
                calculation.</p>
              <a name="Usage" id="Usage"></a>
              <h3>USAGE: MaxExpect &lt;input file&gt; &lt;ct file&gt; [options]</h3>
              <h3 class="leftHeader">Required parameters:</h3>
              <table>
                <tbody>
                  <tr>
                    <td class="cell">&lt;input file&gt;</td>
                    <td> The name of a file containing input data. This input
                      data can be in one of two formats:
                      <ol class="noListTopMargin">
                        <li>Partition function save file (holds base pairing
                          probability data for all pairs and can be generated
                          using the <a href="partition.html">partition</a>
                          interface).</li>
                        <li> <a href="File_Formats.html#Sequence">Sequence file</a>
                          (holds raw sequence: .seq or .fasta). <br>
                          Note that lowercase nucleotides are forced
                          single-stranded in structure prediction. <br>
                          Note that in order to use a squence file, the
                          "sequence" flag must be specified (see "--sequence"
                          below). </li>
                      </ol>
                    </td>
                  </tr>
                  <tr>
                    <td class="cell">&lt;ct file&gt;</td>
                    <td>The name of a <a href="File_Formats.html#CT">CT file</a>
                      to which output will be written.</td>
                  </tr>
                </tbody>
              </table>
              <h3 class="leftHeader">Options that do not require added values:</h3>
              <table>
                <tbody>
                  <tr>
                    <td class="cell">-d, -D, --DNA</td>
                    <td> This flag only matters if the input file is a squence
                      file and has been identified as such (see "--sequence"
                      below). <br>
                      Specify that the sequence is DNA, and DNA parameters are
                      to be used. <br>
                      Default is to use RNA parameters. </td>
                  </tr>
                  <tr>
                    <td class="cell">-h, -H, --help</td>
                    <td>Display the usage details message.</td>
                  </tr>
                  <tr>
                    <td class="cell">--sequence</td>
                    <td>Identify the input file format as a sequence file.</td>
                  </tr>
                </tbody>
              </table>
              <h3 class="leftHeader">Options that require added values:</h3>
              <table>
                <tbody>
                  <tr>
                    <td class="cell">-a, -A, --alphabet</td>
                    <td>Specify the name of a folding alphabet and associated
                      nearest neighbor parameters. The alphabet is the prefix
                      for the thermodynamic parameter files, e.g. "rna" for RNA
                      parameters or "dna" for DNA parameters or a custom
                      extended/modified alphabet. The thermodynamic parameters
                      need to reside in the at the location indicated by
                      environment variable DATAPATH.<br>
                      The default is "rna" (i.e. use RNA parameters). This
                      option overrides the --DNA flag.</td>
                  </tr>
                  <tr>
                  </tr>
                  <tr>
                    <td class="cell">-g, -G, --gamma</td>
                    <td> Specify the weight which should be placed on base
                      pairs. <br>
                      Default is 1.0. </td>
                  </tr>
                  <tr>
                    <td class="cell">-p, -P, --percent</td>
                    <td> Specify a maximum percent difference in folding free
                      energy change in suboptimal structures. Note that
                      suboptimal structures are generated until either the
                      maximum percent free energy difference is reached or until
                      the maximum number of structures is reached (below). <br>
                      Default is 50 percent (specified as 50, not 0.5). </td>
                  </tr>
                  <tr>
                    <td class="cell">-s, -S, --structures</td>
                    <td> Specify a maximum number of structures. Note that
                      suboptimal structures are generated until either the
                      maximum number of structures are reached or the maximum
                      percent difference is reached (above). <br>
                      Default is 1000 structures. </td>
                  </tr>
                  <tr>
                    <td class="cell">-w, -W, --window</td>
                    <td> Specify a window size. Smaller windows (down to zero)
                      will result in a larger set of structures that are
                      similar. Larger windows result in fewer structures that
                      are more different in predicted pairs. <br>
                      Default is 5. </td>
                  </tr>
                </tbody>
              </table>
              <h3 class="leftHeader">Note:</h3>
              <p>MaxExpect scores a structure as the sum of pair probabilities
                for all pairs, multiplied by 2Ã—gamma, plus the sum of being
                unpaired for all unpaired nucleotides (see Lu et al.). This
                score is recorded as an "ENERGY" in CT files, and drawing
                programs display this as the energy of a structure. If a  folding free energy is
                needed, <a href="efn2.html">efn2</a> can be used to calculated
                it using the nearest neighbor parameters.              </p>
              <h3 class="leftHeader"><a name="Refs">References:</a></h3>
              <ol class="noListTopMargin">
                <li> Reuter, J.S. and Mathews, D.H. <br>
                  "RNAstructure: software for RNA secondary structure prediction
                  and analysis." <br>
                  <i>BMC Bioinformatics</i>, 11:129. (2010). </li>
                <li> Lu, Z.J., Gloor, J.W. and Mathews, D.H. <br>
                  "Improved RNA Secondary Structure Prediction by Maximizing
                  Expected Pair Accuracy." <br>
                  <i>RNA</i>, 15:1805-1813. (2009). </li>
              </ol>
            </td>
          </tr>
          <tr>
            <td class="footer" colspan="3">
              <p>Visit <a href="http://rna.urmc.rochester.edu/RNAstructure.html">The
                  Mathews Lab RNAstructure Page</a> for updates and latest
                information.</p>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>
</html>
