<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">
<html>
	<head>
		<title>RNAstructure GUI Help -- References</title>
		<meta http-equiv="Content-Type" content="text/html;charset=utf-8">
		<link href="Stylesheet.css" rel="stylesheet" type="text/css">
	</head>
	<body>
		<div class="bodydiv">
			<table cellpadding="0" cellspacing="0" border="0" class="mainTable">
				<tr class="headerRow">
					<td class="headerBox"><img src="../images/icon.jpg" alt="RNAstructure logo"></td>
					<td class="headerLabel"><h2>RNAstructure GUI Help<br>References</h2></td>
					<td class="headerBox2">
						<ul id="navigator">
							<li><h3 class="noMarginOrPad"><a href="Contents.html">Contents</a></h3></li>
							<li><h3 class="noMarginOrPad"><a href="Index.html">Index</a></h3></li>
						</ul>
					</td>
				</tr>
				<tr>
					<td class="mainContent" colspan="3">
						<div>
							<p>
								<b>Research benefitting from RNAstructure should cite:</b><br>
								J.S. Reuter and D.H. Mathews.<br>
								&quot;RNAstructure: software for RNA secondary structure prediction and analysis.&quot;<br>
								<i>BMC Bioinformatics</i>, 11:129. (2010).
							</p>
							<hr>
						</div>

						<div>
							<p>The algorithms central to RNAstructure are based on the following sources. Research benefitting from these capabilities encapsulated within RNAstructure should cite the appropriate sources.</p>

							<p>
								<b>Most current folding parameters:</b><br>
								D.H. Mathews, M.D. Disney, J.L. Childs, S.J. Schroeder, M. Zuker, and D.H. Turner.<br>
								&quot;Incorporating Chemical Modification Constraints into a Dynamic Programming Algorithm for Prediction of RNA Secondary Structure.&quot;<br>
								<i>Proceedings of the National Academy of Sciences USA</i>, 101:7287-7292. (2004).
							</p>

							<p>
								<b>Previous reference to the folding algorithm, with details on thermodynamic parameter derivation:</b><br>
								D.H. Mathews, J. Sabina, M. Zuker, and D.H. Turner.<br>
								&quot;Expanded Sequence Dependence of Thermodynamic Parameters Improves Prediction of RNA Secondary Structure.&quot;<br>
								<i>Journal of Molecular Biology</i>, 288:911-940. (1999).
							</p>

							<p>
								<b>Prediction of secondary structures at temperatures other than 37Â°C:</b><br>
								Z.J. Lu, D.H. Turner, and D.H. Mathews.<br>
								&quot;A Set of Nearest Neighbor Parameters for Predicting the Enthalpy Change of RNA Secondary Structure Formation.&quot;<br>
								<i>Nucleic Acids Research</i>, 34(17):4912-24. (2006).
							</p>

							<p>
								<b>Partition function for calculating base pair probabilities:</b><br>
								D.H. Mathews.<br>
								&quot;Using an RNA Secondary Structure Partition Function to Determine Confidence in Base Pairs Predicted by Free Energy Minimization.&quot;<br>
								<i>RNA</i>, 10:1178-1190. (2004).
							</p>

							<p>
								<b>Dynalign:</b><br>
								D.H. Mathews.<br>
								&quot;Predicting a Set of Minimal Free Energy Secondary Structures Common to Two Sequences.&quot;<br>
								<i>Bioinformatics</i>, 21:2246-2253. (2005).
							</p>

							<p>
								<b>Extensive use of Dynalign (for extensive acceleration of the algorithm):</b><br>
								A. Harmanci, G. Sharna, and D.H. Mathews.<br>
								&quot;Efficient Pairwise RNA Structure Prediction Using Probabilistic Alignment Constraints in Dynalign.&quot;<br>
								<i>BMC Bioinformatics</i>, 8:130-150. (2007).
							</p>

							<p>
								<b>Previous reference to Dynalign:</b><br>
								D.H. Mathews and D.H. Turner.<br>
								&quot;Dynalign: An Algorithm for Finding the Secondary Structure Common to Two RNA Sequences.&quot;<br>
								<i>Journal of Molecular Biology</i>, 317:191-203. (2002).
							</p>

							<p>
								<b>OligoScreen:</b><br>
								O.V. Mateeva, D.H. Mathews, A.D. Tsodikov, S.A. Shabalina, R.F. Gesteland, J.F. Atkins, and S.M. Freier.<br>
								&quot;Thermodynamic Criteria for High Hit Rate Antisense Oligonucleotide Design.&quot;<br>
								<i>Nucleic Acids Research</i>, 31:4989-4994. (2003).
							</p>

							<p>
								<b>OligoWalk:</b><br>
								D.H. Mathews, M.E. Burkard, S.M. Freier, J.R. Wyatt, and D.H. Turner.<br>
								&quot;Predicting Oligonucleotide Affinity to Nucleic Acid Targets.&quot;<br>
								<i>RNA</i>, 5:1458-1469. (1999).
							</p>

							<p>
								<b>DNA/RNA duplex parameters for DNA oligomers:</b><br>
								N. Sugimoto, S. Nakano, A. Katoh, H. Nakamura, T. Ohmichi, M. Yoneyama, and M. Sasaki.<br>
								&quot;Thermodynamic Parameters to Predict the Stability of RNA/DNA Hybrid Duplexes.&quot;<br>
								<i>Biochemistry</i>, 34:11211-11216. (1995).
							</p>

							<p>
								<b>Prediction of all suboptimal secondary structures:</b><br>
								S. Dian, D.H. Mathews, and D.H. Turner.<br>
								&quot;Interpreting Oligonucleotide Microarray Data to Determine RNA Secondary Structure. Application to the 3' End of Bombyx mori R2 DNA.&quot;<br>
								<i>Biochemistry</i>, 45:9819-9832. (2006).
							</p>

							<p>
								<b>Methods for predicting suboptimal secondary structures:</b><br>
								D.H. Mathews.<br>
								&quot;Revolutions in RNA Secondary Structure Prediction.&quot;<br>
								<i>Journal of Molecular Biology</i>, 359:526-532. (2006).
							</p>

							<p>
								<b>Original reference to generating all suboptimal secondary structures for a sequence:</b><br>
								S. Wuchty, W. Fontant, I,L. Hofacker, and P. Schuster.<br>
								&quot;Complete Suboptimal Folding of RNA and the Stability of Secondary Structures.&quot;<br>
								<i>Biopolymers</i>, 49:145-165. (1999).
							</p>

							<p>
								<b>Stochastic sampling:</b><br>
								Y. Ding and C.E. Lawrence.<br>
								&quot;A Statistical Sampling Algorithm for RNA Secondary Structure Prediction.&quot;<br>
								<i>Nucleic Acids Research</i>, 31:7280-7301. (2003).
							</p>

							<p>
								<b>Using SHAPE data to constrain structure prediction:</b><br>
								K.E. Deigan, T.W. Li, D.H. Mathews, and K.M. Weeks.<br>
								&quot;Accurate SHAPE-directed RNA Structure Determination.&quot;<br>
								<i>Proceedings of the National Academy of Sciences USA</i>, 106: 97-102. (2009).
							</p>

							<p>
								<b>Photocleavage of uracil in the presence of FMN:</b><br>
								P. Burgstaller, T. Hermann, C. Huber, E. Westhof, and M. Famulok.<br>
								&quot;Isoalloxazine derivatives promote photocleavage of natural RNAs at G.U base pairs embedded within helices.&quot;<br>
								<i>Nucleic Acids Ressearch</i>, 25:4018-4027. (1997).
							</p>

							<p>
								<b>ProbKnot pseudoknot prediction algorithm:</b><br>
								S. Bellaousov and D. H. Mathews.<br>
								&quot;ProbKnot: fast prediction of RNA secondary structure including pseudoknots.&quot;<br>
								<i>RNA</i>, 16:1870-1880. (2010).
							</p>

							<p>
								<b>Multilign:</b><br>
								Z. Xu and D. H. Mathews.<br>
&quot;Multilign: An algorithm to predict secondary structures conserved in multiple RNA sequences.&quot;<br>
<em>Bioinformatics</em>, 27:626-632. (2011)						  </p>
						  <p><strong>TurboFold:</strong></p>
						  <p><span class="gapListItem">A. O. Harmanci, H. Sharma,and D. H. Mathews<br>
					      &quot;TurboFold: Iterative Probabilistic Estimation of Secondary Structures for Multiple RNA Sequences.&quot; <br>
					      <i>BMC Bioinformatics</i>, 12:108. (2011).</span></p>
                      </div>
					</td>
				</tr>
				<tr>
					<td class="footer" colspan="3">
						<p>Visit <a href="http://rna.urmc.rochester.edu/RNAstructure.html">The Mathews Lab RNAstructure Page</a> for updates and latest information.</p>
					</td>
				</tr>
			</table>
		</div>
	</body>
</html>
